server.js

```diff
@@ -13,8 +14,126 @@ app.get("/", function (request, response) {
   response.send("Hotel booking server.  Ask for /bookings, etc.");
 });
 
-// TODO add your routes and helper functions here
+//reading all bookings
+app.get("/bookings", (req, res) => {
+  res.status(200).send(bookings);
+});
+
+app.post("/bookings", (req, res) => {
+  const requestBody = req.body;
+
+  const newBooking = {
+    id: requestBody.id,
+    title: requestBody.title,
+    firstName: requestBody.firstName,
+    surname: requestBody.surname,
+    email: requestBody.email,
+    roomId: requestBody.roomId,
+    checkInDate: requestBody.checkInDate,
+    checkOutDate: requestBody.checkOutDate,
+  };
+
+  // validation
+
+  const validBooking =
+    (!!requestBody.id || requestBody.id === 0) &&
+    !!requestBody.title &&
+    !!requestBody.firstName &&
+    !!requestBody.surname &&
+    !!requestBody.email &&
+    emailValidation(requestBody.email) &&
+    !!requestBody.roomId &&
+    !!requestBody.checkInDate &&
+    !!requestBody.checkOutDate &&
+    checkInAndOutDatedValidation(
+      requestBody.checkInDate,
+      requestBody.checkOutDate
+    );
+
+  if (validBooking) {
+    bookings.push(newBooking);
+    res.status(200).send(bookings);
+  } else {
+    res.status(400).send("Some information is missing");
+  }
+});
+
+// Read by Id
+app.get("/bookings/:id", (req, res) => {
+  const bookingId = +req.params.id;
+  const searchedBookingById = bookings.filter(
+    (booking) => booking.id === bookingId
+  );
+  res.status(200).send(searchedBookingById);
+});
+
+//Deleting by id
+app.delete("/bookings/:id", (req, res) => {
+  const bookingId = +req.params.id;
+  const bookingIndex = bookings.findIndex(
+    (booking) => booking.id === bookingId
+  );
+
+  bookings.splice(bookingIndex, 1);
+
+  res.status(200).send(bookings);
+});
+
+//Level 3 search by date
+
+// convet date to numbers
+const dateToNumber = (date) => Number(date.replaceAll("-", ""));
```

Comment:
Intresting idea, and as the date format they ask you to use is non standard  (normally years are last e.g. 27/01/2023) should work alright, but this isn't the correct way to do it. You'd want to make both dates into a date object wether the native JavaScript date object or a moment.js one, and then compare it